### ISSUE TITLE
Dependencies error when building tests only on gitlab ci pipeline (non native)

### ISSUE THREAD

**Issue by deodorel** [quarkusio/quarkus#20444]
### Describe the bug

@aloubyansky 

When the gitlab ci pipeline runs the tests i am getting dependency resolutions errors. 
````
Caused by: io.quarkus.bootstrap.BootstrapException: Failed to create the application model for com.mycompany.mypackage:myservice-svc::jar:1.0.0-SNAPSHOT
Caused by: io.quarkus.bootstrap.resolver.maven.BootstrapMavenException: Failed to resolve dependencies for com.mycompany.mypackage:myservice-svc:jar:1.0.0-SNAPSHOT
Caused by: org.eclipse.aether.resolution.DependencyResolutionException: The following artifacts could not be resolved: com.mycompany.mypackage:myservice-api:yaml:1.0.0-SNAPSHOT, com.mycompany.mypackage:myservice-api:json:1.0.0-SNAPSHOT, com.mycompany.mypackage:myservice-api:yaml:async:1.0.0-SNAPSHOT, com.mycompany.mypackage:myservice-api:json:async:1.0.0-SNAPSHOT: Could not find artifact com.mycompany.mypackage:myservice-api:yaml:1.0.0-SNAPSHOT
Caused by: org.eclipse.aether.resolution.ArtifactResolutionException: The following artifacts could not be resolved: com.mycompany.mypackage:myservice-api:yaml:1.0.0-SNAPSHOT, com.mycompany.mypackage:myservice-api:json:1.0.0-SNAPSHOT, com.mycompany.mypackage:myservice-api:yaml:async:1.0.0-SNAPSHOT, com.mycompany.mypackage:myservice-api:json:async:1.0.0-SNAPSHOT: Could not find artifact com.mycompany.mypackage:myservice-api:yaml:1.0.0-SNAPSHOT
Caused by: org.eclipse.aether.transfer.ArtifactNotFoundException: Could not find artifact com.mycompany.mypackage:myservice-api:yaml:1.0.0-SNAPSHOT
132924 [INFO] Running com.mycompany.mypackage.MyFailingTest
````
I can see the artifacts being deployed in the log previously, in the local repo:

````
6991 [INFO] --- maven-install-plugin:2.4:install (default-install) @ myservice-api ---
6999 [INFO] Installing somelocation/myservice/api/target/myservice-api-1.0.0-SNAPSHOT.jar to somelocation/myservice/.m2/repository/namespace/myservice-api/1.0.0-SNAPSHOT/myservice-api-1.0.0-SNAPSHOT.jar
7010 [INFO] Installing somelocation/myservice/api/pom.xml to somelocation/myservice/.m2/repository/namespace/myservice-api/1.0.0-SNAPSHOT/myservice-api-1.0.0-SNAPSHOT.pom
7022 [INFO] Installing somelocation/myservice/api/target/generated-sources/async-openapi/openapi/openapi.yaml to somelocation/myservice/.m2/repository/namespace/myservice-api/1.0.0-SNAPSHOT/myservice-api-1.0.0-SNAPSHOT-async.yaml
7033 [INFO] Installing somelocation/myservice/api/target/generated-sources/async-openapi/openapi.json to somelocation/myservice/.m2/repository/namespace/myservice-api/1.0.0-SNAPSHOT/myservice-api-1.0.0-SNAPSHOT-async.json
7044 [INFO] Installing somelocation/myservice/api/target/generated-sources/sync-openapi/openapi/openapi.yaml to somelocation/myservice/.m2/repository/namespace/myservice-api/1.0.0-SNAPSHOT/myservice-api-1.0.0-SNAPSHOT.yaml
7055 [INFO] Installing somelocation/myservice/api/target/generated-sources/sync-openapi/openapi.json to somelocation/myservice/.m2/repository/namespace/myservice-api/1.0.0-SNAPSHOT/myservice-api-1.0.0-SNAPSHOT.json
````
The gitlab command is:
````
mvn $MAVEN_CLI_OPTS clean install -DinstallAtEnd=true -DdeployAtEnd=true;
MAVEN_CLI_OPTS=--batch-mode --errors --fail-at-end --show-version -Dmaven.repo.local=.m2/repository -DinstallAtEnd=true -DdeployAtEnd=true --settings $MAVEN_SETTINGS_PATH/$MAVEN_SETTINGS.xml"
MAVEN_SETTINGS_PATH: "/maven-tools"
MAVEN_SETTINGS: "jenkins-settings"
````
my pom.xml:
````
        <dependency>
            <groupId>com.mycompany.mypackage</groupId>
            <artifactId>myservice-api</artifactId>
            <version>1.0.0-SNAPSHOT</version>
            <type>yaml</type>
        </dependency>
        <dependency>
            <groupId>com.mycompany.mypackage</groupId>
            <artifactId>myservice-api</artifactId>
            <version>1.0.0-SNAPSHOT</version>
            <type>json</type>
        </dependency>
        <dependency>
            <groupId>com.mycompany.mypackage</groupId>
            <artifactId>myservice-api</artifactId>
            <version>1.0.0-SNAPSHOT</version>
            <classifier>async</classifier>
            <type>yaml</type>
        </dependency>
        <dependency>
            <groupId>com.mycompany.mypackage</groupId>
            <artifactId>myservice-api</artifactId>
            <version>1.0.0-SNAPSHOT</version>
            <classifier>async</classifier>
            <type>json</type>
        </dependency>
````
then
````
  <plugin>
                <artifactId>maven-compiler-plugin</artifactId>
                <version>${compiler-plugin.version}</version>
            </plugin>
            <plugin>
                <artifactId>maven-surefire-plugin</artifactId>
                <version>${surefire-plugin.version}</version>
                <configuration>
                    <systemPropertyVariables>
                        <java.util.logging.manager>org.jboss.logmanager.LogManager</java.util.logging.manager>
                        <maven.home>${maven.home}</maven.home>
                        <maven.settings>${session.request.userSettingsFile.path}</maven.settings>
                    </systemPropertyVariables>
                </configuration>
            </plugin>`
````

My maven settings contain references to private repos. Nothing special there. The dependencies are deployed on the local maven folder for the build. 
Also, on local of course, everything works perfectly. I tried to reproduce the same versions and command lines from gitlab to no avail. 
Another point, if i have no tests to run, the build passes on gitlab without issues. 

What i tried from other issues found on github:
- making sure to use hardcoded versions (no variables) for the dependencies in question.
- Adding the userSettingsFile line as above
- doing first in the pipeline:       
     mvn $MAVEN_CLI_OPTS dependency:resolve;
     mvn $MAVEN_CLI_OPTS compile -DskipTests;
- trying different versions of the surefire plugin (now i use 3.0.0-M5)
-

### Expected behavior

The pipeline should pass as it passes on local. 

### Actual behavior

the build fails with 
````
Caused by: io.quarkus.bootstrap.BootstrapException: Failed to create the application model for com.mycompany.mypackage:myservice-svc::jar:1.0.0-SNAPSHOT
Caused by: io.quarkus.bootstrap.resolver.maven.BootstrapMavenException: Failed to resolve dependencies for com.mycompany.mypackage:myservice-svc:jar:1.0.0-SNAPSHOT
Caused by: org.eclipse.aether.resolution.DependencyResolutionException: The following artifacts could not be resolved: com.mycompany.mypackage:myservice-api:yaml:1.0.0-SNAPSHOT, com.mycompany.mypackage:myservice-api:json:1.0.0-SNAPSHOT, com.mycompany.mypackage:myservice-api:yaml:async:1.0.0-SNAPSHOT, com.mycompany.mypackage:myservice-api:json:async:1.0.0-SNAPSHOT: Could not find artifact com.mycompany.mypackage:myservice-api:yaml:1.0.0-SNAPSHOT
Caused by: org.eclipse.aether.resolution.ArtifactResolutionException: The following artifacts could not be resolved: com.mycompany.mypackage:myservice-api:yaml:1.0.0-SNAPSHOT, com.mycompany.mypackage:myservice-api:json:1.0.0-SNAPSHOT, com.mycompany.mypackage:myservice-api:yaml:async:1.0.0-SNAPSHOT, com.mycompany.mypackage:myservice-api:json:async:1.0.0-SNAPSHOT: Could not find artifact com.mycompany.mypackage:myservice-api:yaml:1.0.0-SNAPSHOT
Caused by: org.eclipse.aether.transfer.ArtifactNotFoundException: Could not find artifact com.mycompany.mypackage:myservice-api:yaml:1.0.0-SNAPSHOT
132924 [INFO] Running com.mycompany.mypackage.MyFailingTest
````

### How to Reproduce?

Add a test and run it on the gitlab ci

### Output of `uname -a` or `ver`

OS name: "linux", version: "5.4.0-1056-aws", arch: "amd64", family: "unix"

### Output of `java -version`

Java version: 11.0.5, vendor: Oracle Corporation, runtime: /usr/local/openjdk-11

### Quarkus version or git rev

2.1.4.Final

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

3.6.2

**Comment by aloubyansky** [comment]
Thanks, I updated your description to have a better formatting.

**Comment by deodorel** [comment]
Thanks, i wasn't sure about which formatting tag to use ...

**Comment by aloubyansky** [comment]
Looking at 
````
mvn $MAVEN_CLI_OPTS clean install -DinstallAtEnd=true -DdeployAtEnd=true;
MAVEN_CLI_OPTS=--batch-mode --errors --fail-at-end --show-version -Dmaven.repo.local=.m2/repository -DinstallAtEnd=true -DdeployAtEnd=true --settings $MAVEN_SETTINGS_PATH/$MAVEN_SETTINGS.xml"
MAVEN_SETTINGS_PATH: "/maven-tools"
MAVEN_SETTINGS: "jenkins-settings"
````
Shouldn't the `MVN_CLI_OPTS` be initialized before the `mvn` command is executed?

**Comment by deodorel** [comment]
> 
> 
> Looking at
> 
> ```
> mvn $MAVEN_CLI_OPTS clean install -DinstallAtEnd=true -DdeployAtEnd=true;
> MAVEN_CLI_OPTS=--batch-mode --errors --fail-at-end --show-version -Dmaven.repo.local=.m2/repository -DinstallAtEnd=true -DdeployAtEnd=true --settings $MAVEN_SETTINGS_PATH/$MAVEN_SETTINGS.xml"
> MAVEN_SETTINGS_PATH: "/maven-tools"
> MAVEN_SETTINGS: "jenkins-settings"
> ```
> 
> Shouldn't the `MVN_CLI_OPTS` be initialized before the `mvn` command is executed?

Yes it is, i just copied wrongly. The variables are in fact initialized in the top area of the gitlab-ci file.

**Comment by aloubyansky** [comment]
Does it work for you building and running the tests locally?

**Comment by deodorel** [comment]
> 
> 
> Does it work for you building and running the tests locally?

Yes, everything passes on local, with exactly the same command lines.

**Comment by deodorel** [comment]
Maybe what is special about this issue is that the dependencies in question are deployed by the build process itself in the local m2 folder, idk.

**Comment by aloubyansky** [comment]
Just clarifying, is this `somelocation/myservice/.m2/repositor` the expected local Maven repo location? It doesn't match `-Dmaven.repo.local=.m2/repository` on the command line.

Could you please try adding
````
<maven.repo.local>${settings.localRepository}</maven.repo.local>
````
to the surefire plugin config?

**Comment by aloubyansky** [comment]
> Maybe what is special about this issue is that the dependencies in question are deployed by the build process itself in the local m2 folder, idk.

I am not sure what makes it special. It's `mvn install`, right?

**Comment by deodorel** [comment]
> <maven.repo.local>${settings.localRepository}</maven.repo.local>

ok trying right now.

**Comment by deodorel** [comment]
> Just clarifying, is this `somelocation/myservice/.m2/repositor` the expected local Maven repo location? It doesn't match `-Dmaven.repo.local=.m2/repository` on the command line.

Yes, i just did some search and replace to hide company names, packages, but it's the same folder.

**Comment by deodorel** [comment]
ok ran the build with that option, same error

**Comment by aloubyansky** [comment]
Thanks for the info so far. Just in case, would you be able to test a custom Quarkus branch to debug the issue?

**Comment by deodorel** [comment]
> 
> 
> Thanks for the info so far. Just in case, would you be able to test a custom Quarkus branch to debug the issue?

Sure.

**Comment by aloubyansky** [comment]
@deodorel would you mind building this branch https://github.com/aloubyansky/quarkus/tree/2.1.4-debug
To make it install faster you can simply run `mvn -Dquickly` (it will clean install skipping tests and other checks).
This will install Quarkus `2.1.5.CR1`, which you could update your app to and run your tests.
Please look for the INFO messages from `BootstrapMavenContext` in the log. It should log the maven command line args, paths to the local repo, user and global settings and the remote repositories used by the Quarkus bootstrap. Please check those and see what doesn't match your expectations and let me know (what you can reveal). Thanks!

**Comment by deodorel** [comment]
Ok i will try

**Comment by deodorel** [comment]
hello. I tried using your branch, the problem is that on the gitlab ci pipeline where the issue manifests, i can't/am not allowed to deploy on our artifactory repos the debug branch you just created for me ... Because it's not helpful to just build it locally and deploy it on my local repo. 
Could you deploy this on some public artifactory like you would do with a normal quarkus build/release?

**Comment by aloubyansky** [comment]
Would building it as part of the CI pipeline be an option?

**Comment by deodorel** [comment]
good idea. I will give it a try. I can basically run any command on it, even install stuff.

**Comment by aloubyansky** [comment]
with `mvn -Dquickly`?

**Comment by deodorel** [comment]
Yeah. How about if i compile and deploy on the pipeline just the maven plugin and use that version? Would that work.

**Comment by aloubyansky** [comment]
That will have to be the plugin plus its dependencies.

**Comment by deodorel** [comment]
Hello, i was finally able to deploy just the maven resolver artifact to our repo and use the 999-SNAPSHOT version that got build, however, while the result is identical, i fail to find your logs that i saw you added in the code ("global settings", "command line", etc). I do see a lot of info lines from maven, but could it be possible that the resolver has another error level for logging? (like warn)

**Comment by deodorel** [comment]
what's with the readMavenReposFromEnv method you added? can i try to add the BOOTSTRAP_MAVEN_REPOS env variable? what should it contain.

**Comment by deodorel** [comment]
maybe this is why i don't see the logs? LogManager error of type WRITE_FAILURE: The delayed handler's queue was overrun and log record(s) were lost. Did you forget to configure logging?

**Comment by aloubyansky** [comment]
Did you forget to checkout the `2.1.4-debug` branch? It should produce the `2.1.5.CR1` version, not 999-SNAPSHOT. See https://github.com/aloubyansky/quarkus/blob/2.1.4-debug/independent-projects/bootstrap/pom.xml#L14
999-SNAPSHOT is the main branch.

**Comment by deodorel** [comment]
> 
> 
> Did you forget to checkout the `2.1.4-debug` branch? It should produce the `2.1.5.CR1` version, not 999-SNAPSHOT. See https://github.com/aloubyansky/quarkus/blob/2.1.4-debug/independent-projects/bootstrap/pom.xml#L14 999-SNAPSHOT is the main branch.

yeah, pretty much. I will rebuild it.

**Comment by deodorel** [comment]
Hello. So i can clearly see that now the new dependency is used;
Downloading from libs-releases: my_url/artifactory/libs-releases/io/quarkus/quarkus-bootstrap-maven-resolver/2.1.5.CR1/quarkus-bootstrap-maven-resolver-2.1.5.CR1.pom
and then followed by its dependencies. 
However still i can't find the specific logs you added. Do i need to enable them somehow?
If you want i can provide you in private the job's log.

**Comment by aloubyansky** [comment]
Unless it's disabled somewhere, it shouldn't require any logging config, it's logged at INFO level of the `io.quarkus.bootstrap.resolver.maven.BootstrapMavenContext` category.
Are you on Zulip (https://quarkusio.zulipchat.com/)?

**Comment by deodorel** [comment]
In the end the issue i had was related to this https://github.com/quarkusio/quarkus/issues/20444
basically we passed the maven.repo.local as a relative path, and in the maven module the surefire plugin was building the path to the local repo as absolute path of the module + relative repo location, which was wrong. I solved it by passing an absolute path to the maven command line switch.
Big up to @aloubyansky  that created a debug branch for me and added logs that enabled me to see where the problem was.
This can be closed however you see fit.
